File: ios/PlaytomicFoundation/promise/PendingPromise.swift
//
//  PendingPromise.swift
//  Anemone SDK
//
//  Created by Angel Luis Garcia on 13/06/2017.
//  Copyright © 2017 Syltek Solutions S.L. All rights reserved.
//

import Foundation
import PromiseKit

public class PendingPromise<T>: Promise<T> {
    let pending: (fulfill: (T) -> Void, reject: (Error) -> Void)

    public required init() {
        let (promise, resolver) = PromiseKit.Promise<T>.pending()
        self.pending = (resolver.fulfill, resolver.reject)
        super.init(promise: promise)
    }

    @available(*, unavailable, renamed: "init()")
    public required init(
        after _: TimeInterval = 0,
        executeInBackground _: Bool = false,
        resolvers _: @escaping (_ fulfill: @escaping (T) -> Void, _ reject: @escaping (Error) -> Void) throws -> Void
    ) {
        fatalError()
    }

    @available(*, unavailable, renamed: "init()")
    public required init(error _: Error) {
        fatalError()
    }

    @available(*, unavailable, renamed: "init()")
    public required init(value _: T) {
        fatalError()
    }

    public func fulfill(_ value: T) {
        pending.fulfill(value)
    }

    public func reject(_ error: Error) {
        pending.reject(error)
    }
}

File: android/app/src/main/java/com/playtomic/test/foundation/promise/PendingPromise.kt
package com.playtomic.foundation.promise

import org.jdeferred.impl.DeferredObject

/**
 * Created by agarcia on 14/06/2017.
 */
class PendingPromise<T> private constructor(val deferred: DeferredObject<T, Exception, Double>) :
    Promise<T>(deferred.promise()) {

    constructor() : this(deferred = DeferredObject<T, Exception, Double>())

    fun fulfill(value: T) {
        if (deferred.isResolved) {
            return
        }
        deferred.resolve(value)
    }

    fun reject(error: Exception) {
        if (deferred.isResolved) {
            return
        }
        deferred.reject(error)
    }
}
